nexus/src/app/instance.rs

```diff
@@ -1002,6 +1004,43 @@ impl super::Nexus {
                      "instance_id" => %id,
                      "runtime_state" => ?new_runtime_state);
 
+        // If the new state has a newer Propolis ID generation than the current
+        // instance state in CRDB, notify interested parties of this change.
+        //
+        // The synchronization rules here are as follows:
+        //
+        // - Sled agents own an instance's runtime state while an instance is
+        //   running on a sled. Each sled agent prevents concurrent conflicting
+        //   Propolis identifier updates from being sent until previous updates
+        //   are processed.
+        // - Operations that can dispatch an instance to a brand-new sled (e.g.
+        //   live migration) can only start if the appropriate instance runtime
+        //   state fields are cleared in CRDB. For example, while a live
+        //   migration is in progress, the instance's `migration_id` field will
+        //   be non-NULL, and a new migration cannot start until it is cleared.
+        //   This routine must notify recipients before writing new records
+        //   back to CRDB so that these "locks" remain held until all
+        //   notifications have been sent. Otherwise, Nexus might allow new
+        //   operations to proceed that will produce system updates that might
+        //   race with this one.
+        // - This work is not done in a saga. The presumption is instead that
+        //   if any of these operations fail, the entire update will fail, and
+        //   sled agent will retry the update.
```

Comment:
There's no need to unwind these changes, or will the unwind == another update comes along with a higher gen to overwrite?

Reply:
Touched up the comment in 1c36e9ce. The general idea is that any partially-applied updates are correct, and they'll remain correct because there's no way to produce a newer configuration until this one is written back. That's the theory, anyway; thinking about this stuff is making me want to get the TLA+ specs out again.